### DISCUSSION TITLE
Error when calling .wait() .retrieve() .store()

### DISCUSSION THREAD

**Discussion by ramziosta** [Cyfrin/foundry-full-course-cu#1376] [General]
I am not sure if it is a dependency issue or a version issue, but I am not getting the same results as the video for lesson 5.

I am getting errors for calling .wait()  and when I call .retrieve() and .store() on contract as well, and I dont have that option as well, after checking with my ide about what function I can call
![Screenshot 2024-02-02 at 9 27 21 PM](https://github.com/Cyfrin/foundry-full-course-f23/assets/76601606/d7b6f73c-fe1f-4280-b1f4-0aaa8d75737c)
![Screenshot 2024-02-02 at 9 27 41 PM](https://github.com/Cyfrin/foundry-full-course-f23/assets/76601606/a85e7531-c56f-4642-8087-462dbaeb27d1)
![Screenshot 2024-02-02 at 9 27 49 PM](https://github.com/Cyfrin/foundry-full-course-f23/assets/76601606/b67b5e0a-b5f8-4adc-bc8c-c01611992b72)
![Screenshot 2024-02-02 at 9 27 55 PM](https://github.com/Cyfrin/foundry-full-course-f23/assets/76601606/899c9a8b-b41d-4351-84d1-6c4c09c2c666)

So far I have followed the video and I checked all the versions of the dependencies and hardhat

![Screenshot 2024-02-02 at 9 22 14 PM](https://github.com/Cyfrin/foundry-full-course-f23/assets/76601606/39bc27e7-724b-4ccf-843e-06a7ceffdf7b)
![Screenshot 2024-02-02 at 9 24 16 PM](https://github.com/Cyfrin/foundry-full-course-f23/assets/76601606/5b23f5be-dcab-4050-8ccc-597d345d0985)

**Comment by ghost** [comment]
I don't use ethersv6 that much, but this is what I think you should do:

Ethers has a wide range of Types and Interfaces. When you deploy the contract it creates a contract of type [BaseContract](https://docs.ethers.org/v6/api/contract/#BaseContract), in the link you can the avaliable members in the class. You can use the `getFunction` member to invoke functions in the contract just pass the name:

To read from the contract you can use the `staticCall` method:

```javascript
const number = await contract.getFunction("retrieve").staticCall()
```

If you want to store a number (create a transaction) then use the `send` method:

```javascript
const storeTxResponse = await contract.getFunction("store").send("7")
const receipt = await storeTxResponse.wait(1)
```

Regarding waiting for the deployment of the contract I see a `waitForDeployment` method, save in a variable what this method returns to see what it is.

**Reply by ramziosta** [reply]
Thanks that worked for part of the code, but I am still getting this error: it has something to do with not finding an initialized array

> 'VM Exception while processing transaction: invalid opcode',

according to what I found online and from GPT explaining it to me

![Screenshot 2024-02-03 at 10 48 13 PM](https://github.com/Cyfrin/foundry-full-course-f23/assets/76601606/61277487-58ed-4939-9067-493873810ac9)
![Screenshot 2024-02-03 at 10 48 25 PM](https://github.com/Cyfrin/foundry-full-course-f23/assets/76601606/c0564805-3c64-49a4-a926-bbef5bfb1ce3)

**Comment by 0xmillysmith** [comment]
🤔
